cmake/checks/check_01_cxx_features.cmake

```diff
@@ -51,552 +37,404 @@ ENDIF()
-CHECK_CXX_SOURCE_COMPILES(
-  "
-  // common C++11 include files
-  #include <array>
-  #include <condition_variable>
-  #include <type_traits>
+IF(NOT DEAL_II_HAVE_CXX14)
+  MESSAGE(FATAL_ERROR
+    "\nThe current version of deal.II requires a compiler with enabled "
+    "C++14 support. Make sure to use a modern enough compiler (GCC version "
+    "5 onwards, Clang version 4 onwards, or Microsoft MS VS 2015 onwards) "
+    "and check that the compiler flag \"-std=\" is either unset, or set to "
+    "at least c++14.\n\n"
+    )
+ENDIF()
 
-  // thread_local storage specification
-  static thread_local std::array<int,3> p;
+_test_cxx17_support()
 
-  // Check the version language macro, but skip MSVC because
-  // MSVC reports 199711 even in MSVC 2017.
-  #if __cplusplus < 201103L && !defined(_MSC_VER) && !defined(__INTEL_COMPILER)
-  #  error \"insufficient support for C++11\"
-  #endif
 
-  int main()
-  {
-    std::condition_variable c;
-    p[0];
-    c.notify_all();
-
-   // type traits functionality
-   constexpr auto m0 = std::is_trivial<double>::value;
-   (void) m0;
-   constexpr auto m1 = std::is_standard_layout<double>::value;
-   (void) m1;
-   constexpr auto m2 = std::is_pod<double>::value;
-   (void) m2;
-  }
-  "
-  DEAL_II_HAVE_CXX11_FEATURES)
+########################################################################
+#                                                                      #
+#                   Check for various C++ features:                    #
+#                                                                      #
+########################################################################
 
-# clang libc++ bug, see https://llvm.org/bugs/show_bug.cgi?id=20084
-CHECK_CXX_SOURCE_COMPILES(
-  "
-  #include <functional>
-  struct A { void foo() const {} };
-  int main() { A a; std::bind(&A::foo,a)(); return 0; }
-  "
-  DEAL_II_HAVE_CXX11_FUNCTIONAL_LLVMBUG20084_OK)
 
 #
-# On Mac OS-X 10.9 with recent gcc compilers in C++11 mode linking to
-# some standard C library functions, notably toupper and tolower, fail
-# due to unresolved references to these functions.
-#
-# Thanks to Denis Davydov for the testcase.
-#
-# Matthias Maier, 2013
+# In the following we have to avoid
 #
-CHECK_CXX_SOURCE_COMPILES(
-  "
-  #include <ctype.h>
-  int main ()
-  {
-    int c = toupper('a');
-    (void) c;
-  }
-  "
-  DEAL_II_HAVE_CXX11_MACOSXC99BUG_OK)
+SET(_werror_flag "")
+IF(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+  ENABLE_IF_SUPPORTED(_werror_flag "/WX /EHsc")
+ELSE()
+  ENABLE_IF_SUPPORTED(_werror_flag "-Werror")
+  ENABLE_IF_SUPPORTED(_werror_flag "-Wno-unused-command-line-argument")
+ENDIF()
+ADD_FLAGS(CMAKE_REQUIRED_FLAGS "${_werror_flag}")
+
+UNSET_IF_CHANGED(CHECK_CXX_FEATURES_FLAGS_SAVED
+  "${CMAKE_REQUIRED_FLAGS}"
+  DEAL_II_HAVE_FP_EXCEPTIONS
+  DEAL_II_HAVE_COMPLEX_OPERATOR_OVERLOADS
+  DEAL_II_HAVE_CXX17_ATTRIBUTE_FALLTHROUGH
+  DEAL_II_HAVE_ATTRIBUTE_FALLTHROUGH
+  DEAL_II_CXX14_CONSTEXPR_BUG_OK
```

Comment:
```suggestion
  DEAL_II_CXX14_CONSTEXPR_BUG
```

Reply:
I commited and removed your suggesiton again after realizing that `DEAL_II_CXX14_CONSTEXPR_BUG_OK` is indeed what the `DEAL_II_CXX_COMPILER_BUG` exports as cached variable.